collect_app/src/main/java/org/odk/collect/android/formmanagement/ServerFormDownloader.java

```diff
@@ -40,19 +43,31 @@
     private final String formsDirPath;
     private final FormMetadataParser formMetadataParser;
 
-    public ServerFormDownloader(FormSource formSource, FormsRepository formsRepository, File cacheDir, String formsDirPath, FormMetadataParser formMetadataParser) {
+    private final Analytics analytics;
+
+    public ServerFormDownloader(FormSource formSource, FormsRepository formsRepository, File cacheDir, String formsDirPath, FormMetadataParser formMetadataParser, Analytics analytics) {
         this.cacheDir = cacheDir;
         this.formsDirPath = formsDirPath;
         this.multiFormDownloader = new MultiFormDownloader(formsRepository, formSource);
         this.formsRepository = formsRepository;
         this.formMetadataParser = formMetadataParser;
+
+        this.analytics = analytics;
     }
 
     @Override
     public void downloadForm(ServerFormDetails form, @Nullable ProgressReporter progressReporter, @Nullable Supplier<Boolean> isCancelled) throws FormDownloadException, InterruptedException {
         Form formOnDevice = formsRepository.get(form.getFormId(), form.getFormVersion());
-        if (formOnDevice != null && formOnDevice.isDeleted()) {
-            formsRepository.restore(formOnDevice.getId());
+        if (formOnDevice != null) {
+            if (formOnDevice.isDeleted()) {
+                formsRepository.restore(formOnDevice.getId());
+            } else if (!(form.getHash().equals(formOnDevice.getMD5Hash()))) {
+                String formIdentifier = formOnDevice.getDisplayName() + " " + formOnDevice.getId();
+                String formIdHash = FileUtils.getMd5Hash(new ByteArrayInputStream(formIdentifier.getBytes()));
+                analytics.logFormEvent(DOWNLOAD_SAME_FORMID_VERSION, formIdHash);
+                throw new FormDownloadException("You've already downloaded a form with the same ID and version but with different contents. " +
```

Comment:
I went back and forth on whether to localize the download message. I think it's uncommon enough that we don't need to. I've added analytics so that we can verify that claim. I also hope that if it is more common than I expect we'll hear from users on the forum and perhaps we'll choose to work on the wording. "please send all data you have collected with the existing form and delete the data and blank form" is admittedly awkward and I'm open to alternatives.

Reply:
I think I'm happy not localizing the method and with the text but I think it'd be good to separate it out as a display concern here. Could we maybe add a type to `FormDownloadException` (or a subtype) that we can use to match on and then display the right message in more UI focused code?